File: languages/python/lz77.py
import sys

P = 1000000007
WINDOW = 512
MIN_MATCH = 3
MAX_MATCH = 255
ALPHA = 6


def lcg(s):
    return (s * 1103515245 + 12345) & 0x7FFFFFFF


def lz77(n):
    data = bytearray(n)
    s = 42
    for i in range(n):
        s = lcg(s)
        data[i] = s % ALPHA

    pos = 0
    h = 0
    while pos < n:
        best_len = 0
        best_dist = 0
        start = pos - WINDOW
        if start < 0:
            start = 0
        cand = pos - 1
        while cand >= start:                              # nearest distance first
            l = 0
            while pos + l < n and l < MAX_MATCH and data[cand + l] == data[pos + l]:
                l += 1
            if l > best_len:                              # strict > : closest wins ties
                best_len = l
                best_dist = pos - cand
            cand -= 1
        if best_len >= MIN_MATCH:
            h = (h * 31 + 1) % P
            h = (h * 31 + best_dist) % P
            h = (h * 31 + best_len) % P
            pos += best_len
        else:
            h = (h * 31 + 0) % P
            h = (h * 31 + data[pos]) % P
            pos += 1
    return h


if __name__ == "__main__":
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 24000
    print(lz77(n))
    print("lz77(%d)" % n)

File: languages/javascript/lz77.js
"use strict";

// lz77: hand-written LZ77 compressor (WINDOW=512, 6-symbol alphabet), brute-force
// longest-match scan (nearest-distance wins ties), greedy parse. See sort-search.js
// for why the LCG uses Math.imul.

const P = 1000000007;
const WINDOW = 512;
const MIN_MATCH = 3;
const MAX_MATCH = 255;
const ALPHA = 6;

function lcg(s) {
  return (Math.imul(s, 1103515245) + 12345) & 0x7fffffff;
}

function lz77(n) {
  const data = new Uint8Array(n);
  let s = 42;
  for (let i = 0; i < n; i++) {
    s = lcg(s);
    data[i] = s % ALPHA;
  }

  let pos = 0;
  let h = 0;
  while (pos < n) {
    let bestLen = 0;
    let bestDist = 0;
    let start = pos - WINDOW;
    if (start < 0) start = 0;
    let cand = pos - 1;
    while (cand >= start) {
      // nearest distance first
      let l = 0;
      while (pos + l < n && l < MAX_MATCH && data[cand + l] === data[pos + l]) {
        l += 1;
      }
      if (l > bestLen) {
        // strict > : closest wins ties
        bestLen = l;
        bestDist = pos - cand;
      }
      cand -= 1;
    }
    if (bestLen >= MIN_MATCH) {
      h = (h * 31 + 1) % P;
      h = (h * 31 + bestDist) % P;
      h = (h * 31 + bestLen) % P;
      pos += bestLen;
    } else {
      h = (h * 31 + 0) % P;
      h = (h * 31 + data[pos]) % P;
      pos += 1;
    }
  }
  return h;
}

function main() {
  const n = process.argv[2] !== undefined ? parseInt(process.argv[2], 10) : 24000;
  console.log(lz77(n));
  console.log(`lz77(${n})`);
}

main();
